File: GTKMAPS/app/src/main/java/com/example/gtk_maps/Place.java
package com.example.gtk_maps;

import androidx.room.Embedded;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;

import org.osmdroid.views.overlay.Marker;

import java.io.Serializable;
@Entity(foreignKeys = @ForeignKey(entity = Search.class, parentColumns = "id", childColumns = "searchId"))
public class Place implements Serializable {

    @PrimaryKey(autoGenerate = true)
    private long id;
    private long searchId;
    @Embedded
    private Coordinates coordinates;
    @Embedded
    private Address address;
    private String name;
    private String category;
    private String cuisine;
    private String openingHours;
    private String charge;

    /*public Place(double lat, double lon, String name,String category, String cuisine, String openingHours, String charge){
        this.coordinates  = new Coordinates(lat,lon);
        this.category = category;
        this.name = name;
        this.cuisine = cuisine;
        this.openingHours = openingHours;
        this.charge =charge;
    }*/

    public Place(){
        this.coordinates = new Coordinates();
        this.address = new Address();
    }

    public Coordinates getCoordinates() {
        return coordinates;
    }

    public void setCoordinates(Coordinates coordinates) {
        this.coordinates = coordinates;
    }
    public void setCoordinatesWithLatLon(double lat, double lon) {
        this.coordinates =new Coordinates(lat,lon);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getCuisine() {
        return cuisine;
    }

    public void setCuisine(String cuisine) {
        this.cuisine = cuisine;
    }

    public String getOpeningHours() {
        return openingHours;
    }

    public void setOpeningHours(String openingHours) {
        this.openingHours = openingHours;
    }

    public String getCharge() {
        return charge;
    }

    public void setCharge(String charge) {
        this.charge = charge;
    }
    public Address getAddress() {
        return address;
    }

    public void setAddress(Address address) {
        this.address = address;
    }


    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getSearchId() {
        return searchId;
    }

    public void setSearchId(long searchId) {
        this.searchId = searchId;
    }
}

File: GTKMAPS/app/src/main/java/com/example/travel_mate/data/Place.kt
package com.example.travel_mate.data

import androidx.room.ColumnInfo
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import com.example.travel_mate.data.Address
import com.example.travel_mate.data.Coordinates
import com.google.firebase.database.Exclude
import java.io.Serializable
import java.util.UUID

/** [Place]
 * The class stores the data of a found place
 * It also serves as an [androidx.room.Entity] object for the [Room] local database
 *
 * It has may properties I'll better not list them.
 * All of them has [com.example.travel_mate.get] and [set] methods too
 */
@Entity(tableName = "places")
data class Place(
    @PrimaryKey(autoGenerate = false) var uUID: String,
    @ColumnInfo(name = "trip_id")
    @Exclude private var tripId: String? = null,
    @ColumnInfo(name = "place_name") private var name: String? = null,
    @ColumnInfo(name = "cuisine") private var cuisine: String? = null,
    @ColumnInfo(name = "opening_hours")private var openingHours: String? = null,
    @ColumnInfo(name = "charge")private var charge: String? = null,
    @Embedded private var address: Address? = null,
    @Embedded private var coordinates: Coordinates = Coordinates(),
    @ColumnInfo(name = "category") private var category: String? = null,
    @ColumnInfo(name = "contained_by_trip") private var containedByTrip: Boolean = false,
    @Exclude
    @Ignore private var containedByRoute: Boolean = false) : Serializable {

    constructor(): this(
        uUID = UUID.randomUUID().toString()
    )

    constructor(place: Place): this() {
        this.uUID = place.uUID
        this.name = place.name
        this.charge = place.charge
        this.address = place.address
        this.cuisine = place.cuisine
        this.category = place.category
        this.containedByTrip = place.containedByTrip
        this.containedByRoute = place.containedByRoute
        this.coordinates = place.coordinates
        this.openingHours = place.openingHours
    }

    constructor(
        uuid: String,
        name: String,
        charge: String,
        address: Address,
        cuisine: String,
        category: String,
        containedByTrip: Boolean,
        containedByRoute: Boolean,
        coordinates: Coordinates,
        openingHours: String
    ): this() {
        this.uUID = uuid
        this.name = name
        this.charge = charge
        this.address = address
        this.cuisine = cuisine
        this.category = category
        this.containedByTrip = containedByTrip
        this.containedByRoute = containedByRoute
        this.coordinates = coordinates
        this.openingHours = openingHours
    }
    fun getTripId(): String? {
        return this.tripId
    }
    fun setTripId(tripId: String?) {

        this.tripId = tripId
    }


    /*fun getUUID(): String {
        return this.uUID
    }
    fun setUUId(uuid: String) {

        this.uUID = uuid
    }*/
    fun setUUId() {
        this.uUID = UUID.randomUUID().toString()
    }
    fun name(name: String): Place{
        return this.copy( name = name)
    }
    fun setName(name: String){
        this.name = name
    }
    fun getName(): String?{
        return this.name
    }
    fun cuisine(cuisine: String?): Place{
        return this.copy( cuisine = cuisine)
    }
    fun setCuisine(cuisine: String?){
        this.cuisine = cuisine
    }
    fun getCuisine(): String?{
        return this.cuisine
    }
    fun openingHours(openingHours: String?): Place{
        return this.copy(openingHours = openingHours)
    }
    fun setOpeningHours(openingHours: String?){
        this.openingHours = openingHours
    }
    fun getOpeningHours(): String?{
        return this.openingHours
    }
    fun charge(charge: String?): Place{
        return this.copy(charge = charge)
    }
    fun setCharge(charge: String?){
        this.charge = charge
    }
    fun getCharge(): String?{
        return this.charge
    }
    fun category(category: String?): Place{
        return this.copy(category = category)
    }
    fun setCategory(category: String?){
        this.category = category
    }
    fun getCategory(): String?{
        return this.category
    }
    fun address(address: Address?): Place{
        return this.copy(address = address)
    }
    fun setAddress(address: Address?){
        this.address = address
    }
    fun getAddress(): Address?{
        return this.address
    }
    fun coordinates(coordinates: Coordinates): Place{
        return this.copy(coordinates = coordinates)
    }
    fun setCoordinates(coordinates: Coordinates){
        this.coordinates = coordinates
    }
    fun getCoordinates(): Coordinates {
        return this.coordinates
    }
    fun containedByTrip(contained: Boolean): Place {
        return this.copy(containedByTrip = contained)
    }
    fun setContainedByTrip(contained: Boolean) {
        this.containedByTrip = contained
    }
    fun isContainedByTrip(): Boolean {
        return this.containedByTrip
    }
    fun containedByRoute(contained: Boolean): Place {

        return this.copy(containedByRoute = contained)
    }
    fun setContainedByRoute(contained: Boolean) {
        this.containedByRoute = contained
    }

    fun isContainedByRoute(): Boolean {
        return this.containedByRoute
    }

}